File: Data Structures & Algorithms/valid-sudoku/submission-2.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for (int i = 0; i < 9; ++i) {
            set<char> set;
            for (int j = 0; j < 9; ++j) {
                if (set.count(board[i][j]) > 0 && board[i][j] != '.') {
                    return false;
                } else {
                    set.insert(board[i][j]);
                }
            }
        }
        for (int i = 0; i < 9; ++i) {
            set<char> set;
            for (int j = 0; j < 9; ++j) {
                if (set.count(board[j][i]) > 0 && board[j][i] != '.') {
                    return false;
                } else {
                    set.insert(board[j][i]);
                }
            }
        }
        for (int i = 1; i <= 3; ++i) {
            for (int j = 1; j <= 3; ++j) {
                set<char> set;
                for (int a = 0 + 3*(i -1); a < 3 * i ; ++a) {
                    for (int b = 0 + 3*(j - 1); b < 3 * j; ++b) {
                        if (set.count(board[a][b]) > 0 && board[a][b] != '.') {
                            return false;
                        } else {
                            set.insert(board[a][b]);
                        }
                    }
                }
            }
        }
        return true;
    }
};

File: Data Structures & Algorithms/search-for-word-ii/submission-2.py
# buid a trie out of the list of words
# dfs on each element of the grid, keeping track of visited nodes
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        head = TrieNode()
        for word in words:
            curr = head
            for c in word:
                if c not in curr.children:
                    curr.children[c] = TrieNode()
                curr = curr.children[c]
            curr.word = word

        rows = len(board)
        cols = len(board[0])
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        visited = set()
        # loop through each element in board, if in trie lvl 1, dfs through each adjacent element and checking
        # if in child of current trie node, keeping track of visited board elements and current word array
        # .join word array and append to res if full word returns
        visited = set()
        res = []
        def dfs(r, c, node):
            if r < 0 or r >= rows or c < 0 or c >= cols or (r, c) in visited:
                return
            char = board[r][c]
            if char not in node.children:
                return
            nxt = node.children[char]
            if nxt.word:
                res.append(nxt.word)
                nxt.word = None
            visited.add((r, c))
            for dr, dc in dirs:
                dfs(r + dr, c + dc, nxt)
            visited.remove((r, c))

        for r in range(rows):
            for c in range(cols):
                dfs(r, c, head)
        return res
            
            











            
        